src/main/java/org/broadinstitute/hellbender/engine/AssemblyRegionWalker.java

```diff
@@ -0,0 +1,250 @@
+        }
+
+        if ( minAssemblyRegionSize > maxAssemblyRegionSize ) {
+            throw new UserException.BadArgumentValue("--minAssemblyRegionSize", "minAssemblyRegionSize must be <= maxAssemblyRegionSize");
+        }
+
+        if ( maxAssemblyRegionSize > readShardSize ) {
+            throw new UserException.BadArgumentValue("--maxAssemblyRegionSize", "maxAssemblyRegionSize must be <= readShardSize");
+        }
+
+        if ( assemblyRegionPadding > readShardPadding ) {
+            throw new UserException.BadArgumentValue("--assemblyRegionPadding", "assemblyRegionPadding must be <= readShardPadding");
+        }
+
+        final List<SimpleInterval> intervals = hasIntervals() ? intervalsForTraversal : IntervalUtils.getAllIntervalsForReference(getHeaderForReads().getSequenceDictionary());
+        readShards = makeReadShards(intervals);
+    }
+
+    /**
+     * Shard our intervals for traversal into ReadShards using the {@link #readShardSize} and {@link #readShardPadding} arguments
+     *
+     * @param intervals unmodified intervals for traversal
+     * @return List of {@link ReadShard} objects, sharded and padded as necessary
+     */
+    private List<ReadShard> makeReadShards( final List<SimpleInterval> intervals ) {
+        final List<ReadShard> shards = new ArrayList<>();
+
+        for ( final SimpleInterval interval : intervals ) {
+            shards.addAll(ReadShard.shardInterval(interval, readShardSize, readShardPadding, reads, getHeaderForReads().getSequenceDictionary()));
+        }
+
+        return shards;
+    }
+
+    /**
+     * Returns the read filter (simple or composite) that will be applied to the reads.
+     *
+     * The default implementation uses the {@link org.broadinstitute.hellbender.engine.filters.WellformedReadFilter} filter with all default options,
+     * as well as the {@link org.broadinstitute.hellbender.engine.filters.ReadFilterLibrary#MAPPED} filter.
+     *
+     * Default implementation of {@link #traverse()} calls this method once before iterating
+     * over the reads and reuses the filter object to avoid object allocation. Nevertheless, keeping state in filter objects is strongly discouraged.
+     *
+     * Subclasses can override to provide their own filters
+     * Multiple filters can be composed by using {@link org.broadinstitute.hellbender.engine.filters.ReadFilter} composition methods.
+     */
+    public CountingReadFilter makeReadFilter(){
+        return new CountingReadFilter("Wellformed", new WellformedReadFilter(getHeaderForReads()))
+                .and(new CountingReadFilter("Mapped", ReadFilterLibrary.MAPPED));
+    }
+
+    /**
+     * @return The boundaries of the read shard we're currently operating within (ignoring any padding).
+     */
+    public SimpleInterval getCurrentReadShardBounds() {
+        return currentReadShard.getInterval();
+    }
+
+    @Override
+    public void traverse() {
```

Comment:
do you want to keep `traverse` non-final? If yes, document what the subclasses should do when they override it. At least say if they are required to call super or not.